src/exchanges/crdtExchange.ts

```diff
@@ -0,0 +1,577 @@
+import type { Client, CombinedError, Exchange, Operation, OperationResult, RequestPolicy } from 'urql';
+
+import {
+  createRequest,
+  makeOperation,
+  getOperationName,
+  makeResult,
+} from '@urql/core';
+import { enableMapSet, produce } from 'immer';
+import { filter, fromArray, interval, makeSubject, merge, mergeMap, pipe, share, subscribe, tap } from 'wonka';
+
+import { get as lGet, isArray, values, keyBy, mergeWith, set as lSet } from 'lodash';
+
+enableMapSet();
+
+const toRequestPolicy = (
+  operation: Operation,
+  requestPolicy: RequestPolicy
+): Operation => {
+  return makeOperation(operation.kind, operation, {
+    ...operation.context,
+    requestPolicy,
+  });
+};
+
+export const isOfflineError = (error: undefined | CombinedError) =>
+  error &&
+  error.networkError &&
+  !error.response &&
+  ((typeof navigator !== 'undefined' && navigator.onLine === false) ||
+    /request failed|failed to fetch|network\s?error/i.test(
+      error.networkError.message
+    ));
+
+export const isDeadlockMutation = (error: undefined | CombinedError) => {
+  const deadlockRegex = /deadlock/i;
+  return error && error.graphQLErrors.some(e=> e.message.match(deadlockRegex));
+};
+
+export type CrdtExchangeOpts =  {
+  isRetryableError: (res: OperationResult) => boolean;
+}
+
+/**
+ * Returns the value associated with the *more recent* timestamp. Null is considered older than any timestamp.
+ *
+ * Returns undefined when both timestamps are falsy.
+ *
+ * @param ts1 timestamp associated with value 1
+ * @param ts2 timestamp associated with value 2
+ * @param value1 value 1
+ * @param value2 value 2
+ */
+ const compareTs = (ts1: string | null, ts2: string | null, value1: any, value2: any) => {
+  if (!ts1 && !ts2) {
+    return;
+  }
+  if (!ts1) {
+    return value2;
+  }
+  if (!ts2) {
+    return value1;
+  }
+  const compareRes = ts1.localeCompare(ts2);
+  if (compareRes <= 0) {
+    return value2;
+  }
+  return value1;
+}
+
+const mergeWithTimestampsCustomizer = (targetValue: any, srcValue: any, key: string, target: any, source: any): any => {
+  // TODO: way to standardize uuid key
+  if (isArray(targetValue)) {
+    return values(mergeWith(keyBy(targetValue, 'uuid'), keyBy(srcValue, 'uuid'), mergeWithTimestampsCustomizer));
+  }
+
+  if (target?.timestamps?.[key] || source?.timestamps?.[key]) {
+    return compareTs(target?.timestamps?.[key], source?.timestamps?.[key], targetValue, srcValue);
+  }
+
+  // TODO: way to standardize timestamps key
+  if (key === 'timestamps') {
+    return mergeWith({}, targetValue, srcValue, (v, srcV) => {
+      return compareTs(v, srcV, v, srcV);
+    });
+  }
+}
+
+const mergeWithTimestamps = (existing: any, newValues: any) => {
+  return mergeWith({}, existing, newValues, mergeWithTimestampsCustomizer);
+}
+
+// TODO: use config information to infer this
+const isCrdtOperation = (operation: Operation) => {
+  const operationName = getOperationName(operation.query)!;
```

Comment:
Yes, that's correct. `includes(undefined)` will always be false (unless your array also had the value `undefined`). But TS believes `includes` should only accept `string`. It may be more clear to instead do `includes(operationName!)`, but it's the same either way.